plugins/elantp/fu-elantp-hid-haptic-device.c

```diff
@@ -0,0 +1,865 @@
+/*
+ * Copyright (C) 2020 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+#include <stdio.h>
+#include <linux/hidraw.h>
+#include <linux/input.h>
+
+#include "fu-elantp-common.h"
+#include "fu-elanhaptic-firmware.h"
+#include "fu-elantp-hid-haptic-device.h"
+
+struct _FuElantpHidHapticDevice {
+	FuUdevDevice parent_instance;
+	guint16 ic_page_count;
+	guint16 iap_type;
+	guint16 tp_iap_ctrl;
+	guint16 iap_ctrl;
+	guint16 iap_password;
+	guint16 module_id;
+	guint16 fw_page_size;
+	guint8 pattern;
+	gint16 driver_ic;
+	guint8 iap_ver;
+};
+
+G_DEFINE_TYPE(FuElantpHidHapticDevice, fu_elantp_hid_haptic_device, FU_TYPE_UDEV_DEVICE)
+
+#define FU_ELANTP_DEVICE_IOCTL_TIMEOUT 5000 /* ms */
+
+static FuElantpHidDevice *
+fu_elantp_haptic_device_get_parent(FuDevice *self, GError **error)
+{
+	FuDevice *parent = fu_device_get_parent(FU_DEVICE(self));
+	if (parent == NULL) {
+		g_set_error_literal(error, FWUPD_ERROR, FWUPD_ERROR_INTERNAL, "no parent set");
+		return NULL;
+	}
+	return FU_ELANTP_HID_DEVICE(FU_UDEV_DEVICE(parent));
+}
+
+static gboolean
+fu_elantp_hid_haptic_device_detach(FuDevice *device, FuProgress *progress, GError **error);
+
+static void
+fu_elantp_hid_haptic_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuElantpHidHapticDevice *self = FU_ELANTP_HID_HAPTIC_DEVICE(device);
+	fu_string_append_kx(str, idt, "ModuleId", self->module_id);
+	fu_string_append_kx(str, idt, "Pattern", self->pattern);
+	fu_string_append_kx(str, idt, "FwPageSize", self->fw_page_size);
+	fu_string_append_kx(str, idt, "IcPageCount", self->ic_page_count);
+	fu_string_append_kx(str, idt, "IapType", self->iap_type);
+	fu_string_append_kx(str, idt, "TpIapCtrl", self->tp_iap_ctrl);
+	fu_string_append_kx(str, idt, "IapCtrl", self->iap_ctrl);
+	fu_string_append_kx(str, idt, "DriverIC", self->driver_ic);
+	fu_string_append_kx(str, idt, "IAPVersion", self->iap_ver);
+}
+
+static gboolean
+fu_elantp_hid_haptic_device_probe(FuDevice *device, GError **error)
+{
+	/* set the physical ID */
+	return fu_udev_device_set_physical_id(FU_UDEV_DEVICE(device), "hid", error);
+}
+
+static gboolean
+fu_elantp_hid_haptic_device_send_cmd(FuDevice *self,
+			      guint8 *tx,
```

Comment:
const perhaps?

Reply:
what is mean "const perhaps?"

Reply:
e.g. `const guint8 *tx`

Reply:
This will appear WARNING
../plugins/elantp/fu-elantp-hid-haptic-device.c: In function ‘fu_elantp_hid_haptic_device_send_cmd’:
../plugins/elantp/fu-elantp-hid-haptic-device.c:89:35: warning: passing argument 3 of ‘fu_udev_device_ioctl’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   89 |                                   tx,